File: services/scheduler-engine/src/scheduler_engine/schemas/schedule_generate.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field


class ScheduleDayAssignment(BaseModel):
    date: str = Field(pattern=r"^\d{4}-\d{2}-\d{2}$")
    shift_template_id: str = Field(alias="shiftTemplateId")

    model_config = {"populate_by_name": True}


class ShiftDefinition(BaseModel):
    role: Literal["worker", "boss"]
    required_workers: int = Field(alias="requiredWorkers", ge=1)
    start: str
    end: str
    weekdays: list[str]

    model_config = {"populate_by_name": True}


class ShiftTemplate(BaseModel):
    id: str
    name: str
    shifts: list[ShiftDefinition]
    created_at: str = Field(alias="createdAt")

    model_config = {"populate_by_name": True}


class Holiday(BaseModel):
    id: str
    created_at: str = Field(alias="createdAt")
    name: str | None
    date: str
    start: int | None
    end: int | None

    model_config = {"populate_by_name": True}


class WorkerPayload(BaseModel):
    id: str
    first_name: str = Field(alias="firstName")
    last_name: str = Field(alias="lastName")
    role: Literal["worker", "boss"]
    priority: int = Field(ge=1, le=10)
    checker: bool
    deleted: bool

    model_config = {"populate_by_name": True}


class WorkerDraftPayload(BaseModel):
    draft_id: str = Field(alias="draftId")
    worker_id: str = Field(alias="workerId")
    file_name: str = Field(alias="fileName")
    content_base64: str = Field(alias="contentBase64")

    model_config = {"populate_by_name": True}


class GenerateScheduleRequest(BaseModel):
    year: int = Field(ge=2000, le=2100)
    month: int = Field(ge=1, le=12)
    day_assignments: list[ScheduleDayAssignment] = Field(alias="dayAssignments", min_length=1)
    holidays: list[Holiday]
    shift_templates: list[ShiftTemplate] = Field(alias="shiftTemplates")
    worker_drafts: list[WorkerDraftPayload] = Field(alias="workerDrafts")
    workers: list[WorkerPayload]

    model_config = {"populate_by_name": True}


class GenerateScheduleResponse(BaseModel):
    job_id: str = Field(alias="jobId")
    status: Literal["completed", "failed"]
    message: str
    draft_count: int = Field(alias="draftCount")
    holiday_count: int = Field(alias="holidayCount")
    worker_count: int = Field(alias="workerCount")
    assignment_count: int = Field(alias="assignmentCount")
    workers: list[dict[str, Any]]
    assignments: list[dict[str, Any]]
    solver_status: Literal["optimal", "feasible", "infeasible"] = Field(alias="solverStatus")
    unassigned_slot_ids: list[str] = Field(alias="unassignedSlotIds")

    model_config = {"populate_by_name": True, "by_alias": True}

File: packages/shared/src/schemas/schedule-generate.ts
import { z } from 'zod';

import { shiftTemplateIdSchema } from './shift-template';

const dateStringSchema = z
  .string()
  .regex(/^\d{4}-\d{2}-\d{2}$/, 'Data musi być w formacie RRRR-MM-DD');

export const scheduleDayAssignmentSchema = z.object({
  date: dateStringSchema,
  shiftTemplateId: shiftTemplateIdSchema,
});

export const schedulePreviewHalfCellSchema = z.object({
  text: z.string().nullable(),
  fill: z.enum(['none', 'yellow', 'purple']),
});

export const schedulePreviewDayCellSchema = z.object({
  start: schedulePreviewHalfCellSchema,
  end: schedulePreviewHalfCellSchema,
});

export const schedulePreviewWorkerSchema = z.object({
  workerId: z.string(),
  firstName: z.string(),
  lastName: z.string(),
  rows: z.array(z.array(schedulePreviewDayCellSchema)),
});

export const schedulePreviewSchema = z.object({
  year: z.number().int(),
  month: z.number().int().min(1).max(12),
  daysInMonth: z.number().int().min(28).max(31),
  weekdays: z.array(z.string()),
  dayNumbers: z.array(z.number().int()),
  workers: z.array(schedulePreviewWorkerSchema),
});

export const generateScheduleInputSchema = z.object({
  year: z.number().int().min(2000).max(2100),
  month: z.number().int().min(1).max(12),
  dayAssignments: z.array(scheduleDayAssignmentSchema).min(1),
});

export const generateScheduleResultSchema = z.object({
  jobId: z.string().min(1),
  year: z.number().int(),
  month: z.number().int().min(1).max(12),
  status: z.enum(['accepted', 'failed']),
  draftCount: z.number().int().nonnegative(),
  holidayCount: z.number().int().nonnegative(),
  assignmentCount: z.number().int().nonnegative(),
  totalSlotCount: z.number().int().nonnegative(),
  solverStatus: z.enum(['optimal', 'feasible', 'infeasible']),
  message: z.string(),
  preview: schedulePreviewSchema,
  unassignedSlotIds: z.array(z.string()),
});

export const exportGrafikPdfInputSchema = z.object({
  preview: schedulePreviewSchema,
});

export const exportGrafikPdfResultSchema = z.object({
  fileName: z.string().min(1),
  contentBase64: z.string().min(1),
});

export type ScheduleDayAssignment = z.infer<typeof scheduleDayAssignmentSchema>;
export type SchedulePreviewHalfCell = z.infer<typeof schedulePreviewHalfCellSchema>;
export type SchedulePreviewDayCell = z.infer<typeof schedulePreviewDayCellSchema>;
export type SchedulePreviewWorker = z.infer<typeof schedulePreviewWorkerSchema>;
export type SchedulePreview = z.infer<typeof schedulePreviewSchema>;
export type GenerateScheduleInput = z.infer<typeof generateScheduleInputSchema>;
export type GenerateScheduleResult = z.infer<typeof generateScheduleResultSchema>;
export type ExportGrafikPdfInput = z.infer<typeof exportGrafikPdfInputSchema>;
export type ExportGrafikPdfResult = z.infer<typeof exportGrafikPdfResultSchema>;
